src/bin/cargo/cli.rs

```diff
@@ -231,26 +231,25 @@ fn expand_aliases(
 ) -> Result<(ArgMatches, GlobalArgs), CliError> {
     if let Some((cmd, args)) = args.subcommand() {
         let exec = commands::builtin_exec(cmd);
-        let aliased_cmd = super::aliased_command(config, cmd).or_else(|e| {
-            // HACK: `cargo version` must always work even with malformed configs.
-            // Here we treat the error as the aliased command not found,
-            // so it falls back to call builtin `cargo version`
-            if cmd == "version" {
-                Ok(None)
-            } else {
-                Err(e)
-            }
-        })?;
+        let aliased_cmd = super::aliased_command(config, cmd);
 
         match (exec, aliased_cmd) {
-            (Some(_), Some(_)) => {
+            (_, Err(e)) => {
+                // HACK: `cargo version` must always work even with malformed configs.
+                // so when cmd is exactly `version`, it ignores the error
+                // and falls back to call builtin `cargo version` then.
+                if cmd != "version" {
+                    return Err(e.into());
+                }
+            }
```

Comment:
I had assumed we would prioritize this below `(Some(_), Err(_)) | (Some(_), Ok(None))`.  My thinking was "If aliasing isn't involved, why report the error?".

What are your thoughts?

Reply:
The branch is not required anymore. Nice suggestion! Thanks a lot.